### PR TITLE
feat: create EPO order in manage orders event (FPLA-2950)

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2950

### Change description ###
Create Emergency protection order in manage orders event.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json [hunk 1]
```diff
@@ -29,12 +29,130 @@
...
     "CaseTypeID": "CARE_SUPERVISION_EPO",
     "CaseEventID": "manageOrders",
-    "CaseFieldID": "manageOrdersFurtherDirections",
+    "CaseFieldID": "manageOrdersEpoType",
     "PageFieldDisplayOrder": 3,
+    "DisplayContext": "MANDATORY",
+    "PageID": "OrderDetails",
+    "PageDisplayOrder": 5,
+    "PageColumnNumber": 1,
+    "FieldShowCondition": "orderTempQuestions.epoOrderType=\"YES\"",
```
Comment (lewisbirks) [inline ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json:38]: I feel like this could have a more generic name, something like just `orderTempQuestions.epo` or something like that so that you don't need to specifiy a new field for each question to be dislpayed when it is an epo.
This field would then replace `epoOrderType`, `epoPreventRemoval`, `epoExpiryDate`, `epoIncludePhrase` and `epoChildrenDescription`

File: ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json [hunk 2]
```diff
@@ -29,12 +29,130 @@
...
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "manageOrders",
+    "CaseFieldID": "manageOrdersEpoRemovalAddress",
+    "PageFieldDisplayOrder": 4,
+    "DisplayContext": "MANDATORY",
+    "PageID": "OrderDetails",
+    "PageDisplayOrder": 5,
+    "PageColumnNumber": 1,
+    "FieldShowCondition": "orderTempQuestions.epoPreventRemoval=\"YES\" AND manageOrdersEpoType=\"PREVENT_REMOVAL\"",
```
Comment (lewisbirks) [inline ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json:51]: not sure the first part of this condition is required

File: e2e/pages/events/manageOrders.page.js [hunk 3]
```diff
@@ -75,6 +116,47 @@ const selectChildren = async (option, indexes = []) => {
...
+  await I.fillDate(date, section4.exclusionStartDate);
+  await I.runAccessibilityTest();
+};
+
+const uploadPowerOfArrest = async (file) => {
+  I.attachFile(section4.powerOfArrest, file);
+  await I.runAccessibilityTest();
+};
+
+const clickPostCodeLink = async (linkLabel) => {
```
Comment (lewisbirks) [inline e2e/pages/events/manageOrders.page.js:149]: can you use the address postcode lookup fragment for this stuff?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersAboutToSubmitControllerTest.java [hunk 4]
```diff
@@ -118,6 +136,46 @@ void shouldBuildNewOrderObject() {
...
+                    .build())
+                .dateTimeIssued(now())
+                .approvalDateTime(manageOrdersEventData.getManageOrdersApprovalDateTime())
+                .childrenDescription("first1 last1, first2 last2")
+                .document(DOCUMENT_PDF_REFERENCE)
+                .unsealedDocumentCopy(DOCUMENT_WORD_REFERENCE)
+                .build()));
+    }
+
+    private ManageOrdersEventData buildManageOrdersEventData() {
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersAboutToSubmitControllerTest.java:163]: I would make this local to the method as it is very specific event data

Comment (lewisbirks) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersAboutToSubmitControllerTest.java:163]: What you could do instead is also extract the event data from `buildCaseData()` and make that local to the other test. As such that method is now building a generic case data for you to then add to.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersAboutToSubmitControllerTest.java [hunk 5]
```diff
@@ -118,6 +136,46 @@ void shouldBuildNewOrderObject() {
...
+            .manageOrdersEpoRemovalAddress(Address.builder().addressLine1("address1").postcode("postcode").build())
+            .manageOrdersChildrenDescription("first1 last1, first2 last2")
+            .manageOrdersFurtherDirections("test directions")
+            .manageOrdersExclusionRequirement("Yes")
+            .manageOrdersWhoIsExcluded("John")
+            .manageOrdersExclusionStartDate(dateNow().plusDays(2))
+            .manageOrdersPowerOfArrest(DOCUMENT_PDF_REFERENCE)
+            .build();
+    }
+
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersAboutToSubmitControllerTest.java:178]: Would be useful to add your fields to the below method, so that we can check that they are removed at the end of the flow

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java [hunk 6]
```diff
@@ -83,15 +94,41 @@ void orderSelectionShouldPopulateQuestionConditionHolder() {
...
+            "epoChildrenDescription", "NO",
+            "epoExpiryDate", "NO",
+            "epoPreventRemoval", "NO"
+        ));
+
+        assertThat(response.getData().get("orderTempQuestions")).isEqualTo(expectedQuestions);
+    }
+
+    @Test
+    void epoOrderSelectionShouldPopulateQuestionConditionHolder() {
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java:112]: Although I am not complaining about it do we need a test for the specific questions of the order type in the integration tests. Surely that would be covered in the unit tests and in the int test we just need to make sure that it has populated it correctly for one type of order to ensure that the data is actually returned

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java [hunk 7]
```diff
@@ -214,11 +252,151 @@ void reviewShouldNotAlterCaseData() {
...
+
     @Test
     void shouldThrowExceptionWhenMidEventUrlParameterDoesNotMatchSectionNames() {
         assertThatThrownBy(() -> postMidEvent(CaseData.builder().build(), "does-not-match"))
             .getRootCause()
             .isInstanceOf(IllegalArgumentException.class)
             .hasMessage("No enum constant uk.gov.hmcts.reform.fpl.model.order.OrderSection.DOES_NOT_MATCH");
     }
+
+    private static Map<String, String> getOrderQuestions() {
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java:357]: Don't think this method should be needed, but if you don't want I think the name should be different and it doesn't need to be static. The name implies that it is all the order questions and then in the methods you keep adding to them.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/OrderQuestionBlock.java [hunk 8]
```diff
@@ -12,14 +12,25 @@
         List.of("judgeAndLegalAdvisor")),
     APPROVAL_DATE("approvalDate", "Approval Date", OrderSection.ISSUING_DETAILS,
         List.of("manageOrdersApprovalDate")),
+    APPROVAL_DATE_TIME("approvalDateTime", "Approval Date Time", OrderSection.ISSUING_DETAILS,
+        List.of("manageOrdersApprovalDateTime")),
     WHICH_CHILDREN("whichChildren", "Which children", OrderSection.CHILDREN_DETAILS,
         List.of("orderAppliesToAllChildren", "children_label", "childSelector")),
+    EPO_TYPE("epoOrderType", "EPO Type", OrderSection.ORDER_DETAILS, List.of("manageOrdersEpoType")),
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/OrderQuestionBlock.java:19]: As mentioned before I think we can treat all epo fields in the issuing details section as one question block

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocmosisParameterGenerator.java [hunk 9]
```diff
@@ -2,14 +2,21 @@
...
+
 public interface DocmosisParameterGenerator {
 
     Order accept();
 
     DocmosisParameters generate(CaseData caseData);
 
     DocmosisTemplates template();
+
+    default List<DocumentReference> additionalDocuments(CaseData caseData) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocmosisParameterGenerator.java:19]: I don't feel like this belongs as part of this interface. This is nothing to do with the DocmosisParameters. Instead it should probs be a new interface that is added to the OrderDocumentGeneratorHolder

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMerger.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class DocumentMerger {
+
+    private final DocumentDownloadService documentDownloadService;
+    private final DocumentConversionService documentConversionService;
+
+    public DocmosisDocument mergeDocuments(
+        DocmosisDocument originalDocument, List<DocumentReference> additionalDocuments) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMerger.java:30]: vv minor but if you have the method declaration like so
```java
    public DocmosisDocument mergeDocuments(DocmosisDocument originalDocument,
                                           List<DocumentReference> additionalDocuments) {
```
it looks better in my opinion

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMerger.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            pdfMergerUtility.mergeDocuments(setupMainMemoryOnly());
+            return new DocmosisDocument(originalDocument.getDocumentTitle(), docOutputStream.toByteArray());
+        } catch (IOException e) {
+            throw new DocumentMergeException(
+                "Exception occurred while merging documents for " + originalDocument.getDocumentTitle(), e);
+        }
+    }
+
+    private List<InputStream> getPdfFilesToMerge(
+        DocmosisDocument originalDocument, List<DocumentReference> additionalDocuments) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMerger.java:52]: see previous

File: ccd-definition/CaseField/CareSupervision/manageOrders/manageOrdersFields.json [hunk 12]
```diff
@@ -54,5 +63,87 @@
...
+    "Label": "Date power of exclusion starts",
+    "SecurityClassification": "PUBLIC",
+    "Searchable": "N"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "manageOrdersPowerOfArrest",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "Label": "Upload power of arrest, if required",
+    "FieldType": "Document",
```
Comment (lewisbirks) [inline ccd-definition/CaseField/CareSupervision/manageOrders/manageOrdersFields.json:145]: Are there any restrictions on the document type here, we are converting to it pdf later so we probs don't want to allow them to upload anything

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/history/SealedOrderHistoryService.java [hunk 13]
```diff
@@ -47,9 +47,11 @@
             .orderType(manageOrdersEventData.getManageOrdersType().name()) // hidden field, to store the type
             .title(manageOrdersEventData.getManageOrdersType().getHistoryTitle())
             .children(selectedChildren)
-            .judgeAndLegalAdvisor(getJudgeForTabView(caseData.getJudgeAndLegalAdvisor(), caseData.getAllocatedJudge()))
+            .judgeAndLegalAdvisor(
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/history/SealedOrderHistoryService.java:50]: unnecessary change

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidator.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    public OrderQuestionBlock accept() {
+        return EPO_EXPIRY_DATE;
+    }
+
+    @Override
+    public List<String> validate(CaseData caseData) {
+        final LocalDateTime epoApprovalTime = caseData.getManageOrdersEventData().getManageOrdersApprovalDateTime();
+        final LocalDateTime epoEndTime = caseData.getManageOrdersEventData().getManageOrdersEndDateTime();
+
+        if (epoEndTime == null || epoApprovalTime == null) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidator.java:42]: Neither of these would be null at this point should they?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidator.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = {@Autowired})
+public class EPOEndDateValidator implements QuestionBlockOrderValidator {
+
+    private static final String INVALID_TIME_MESSAGE = "Enter a valid time";
+    private static final String FUTURE_DATE_MESSAGE = "Enter an end date in the future";
+    private static final String END_DATE_RANGE_MESSAGE = "Emergency protection orders cannot last longer than 8 days";
+
+    private static final LocalTime MIDNIGHT = LocalTime.of(0, 0, 0);
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidator.java:27]: There is already a constant in the LocalTime class for this, called the exact same thing 😆

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPORemovalAddressValidator.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Override
+    public OrderQuestionBlock accept() {
+        return EPO_PREVENT_REMOVAL;
+    }
+
+    @Override
+    public List<String> validate(CaseData caseData) {
+        final ManageOrdersEventData eventData = caseData.getManageOrdersEventData();
+        final Address removalAddress = eventData.getManageOrdersEpoRemovalAddress();
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPORemovalAddressValidator.java:33]: The address is a mandatory field, why are we validating that it is empty or are we not trusting xui?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C23EPODocumentParameterGeneratorTest.java [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    public static final String EXCLUDE_PERSON = "Test user1";
+    public static final Address REMOVAL_ADDRESS = Address.builder()
+        .addressLine1("12 street").postcode("SW1").country("UK").build();
+
+    private final ChildrenService childrenService = mock(ChildrenService.class);
+    private final LocalAuthorityNameLookupConfiguration laNameLookup = mock(
+        LocalAuthorityNameLookupConfiguration.class
+    );
+
+    private final C23EPODocumentParameterGenerator underTest = new C23EPODocumentParameterGenerator(
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C23EPODocumentParameterGeneratorTest.java:57]: If you don't like the long lines here what you can do is annotate the class with `@ExtendWith({MockitoExtension.class})` and then annotate the mocks with `@Mock` and then annotate this field with `@InjectMocks`. With that you can remove all the initialisations if you want

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    private static final String ADDITIONAL_FILENAME = "document2.pdf";
+
+    private static final DocmosisDocument DOCMOSIS_DOCUMENT = DocmosisDocument.builder()
+        .documentTitle(ORIGINAL_FILENAME).bytes(new byte[]{1}).build();
+
+    private static final DocumentReference DOCUMENT_REFERENCE = testDocumentReference(ADDITIONAL_FILENAME);
+
+    @Test
+    void shouldMergeTheDocuments() {
+        final byte[] originalDocument = readBytes("documents/document1.pdf");
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java:44]: Do we need the additional documents? Can we not just have some dummy byte arrays like you do with the additionalDocumentBytes?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+            .thenReturn(additionalDocumentBytes);
+        when(documentConversionService.convertToPdf(eq(additionalDocumentBytes), anyString()))
+            .thenReturn(additionalDocumentBytes);
+
+        Exception exception = assertThrows(DocumentMergeException.class,
+            () -> underTest.mergeDocuments(DOCMOSIS_DOCUMENT, List.of(DOCUMENT_REFERENCE))
+        );
+
+        assertThat(exception.getMessage())
+            .isEqualTo(String.format("Exception occurred while merging documents for %s",
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java:88]: Don't think that you need the String.format here, you know the file name from the decalared fields above.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/EPOAddressPrePopulatorTest.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.util.Map;
+
+class EPOAddressPrePopulatorTest {
+
+    private final EPOAddressPrePopulator underTest = new EPOAddressPrePopulator();
+
+    @Test
+    void testAcceptsEPORemovalAddressSection() {
+        Assertions.assertThat(underTest.accept()).isEqualTo(OrderQuestionBlock.EPO_PREVENT_REMOVAL);
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/prepopulator/question/EPOAddressPrePopulatorTest.java:18]: static import

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPORemovalAddressValidatorTest.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+class EPORemovalAddressValidatorTest {
+
+    private static final String INVALID_ADDRESS_LINE = "Enter a valid address for the contact";
+    private static final String INVALID_POST_CODE = "Enter a postcode for the contact";
+
+    private final EPORemovalAddressValidator underTest = new EPORemovalAddressValidator();
+
+    @Test
+    void accept() {
+        AssertionsForClassTypes.assertThat(underTest.accept()).isEqualTo(EPO_PREVENT_REMOVAL);
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPORemovalAddressValidatorTest.java:24]: Although this class will work fine for assertions it is technically the wrong one. This class is for java 7 and lower and contains work arounds for features that were not imlpemented there but are in java 8+. You already statically import the other assertThat so this one can be removed.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/OrderValidatorHolderTest.java [hunk 22]
```diff
@@ -49,7 +63,10 @@ void blockToValidatorCached() {
         underTest.blockToValidator();
         assertThat(underTest.blockToValidator()).isEqualTo(Map.of(
             APPROVAL_DATE, approvalDateValidator,
-            WHICH_CHILDREN, whichChildrenValidator
+            APPROVAL_DATE_TIME, approvalDateTimeValidator,
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/OrderValidatorHolderTest.java:66]: might be nice to create this map in the setUp method and then just use that new map in both the tests

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/ManageOrderDocumentScopedFieldsCalculatorTest.java [hunk 23]
```diff
@@ -2,32 +2,38 @@
...
-                "childrenDetailsSectionSubHeader",
-                "orderDetailsSectionSubHeader"
-            )
+        assertThat(underTest.calculate()).containsExactlyInAnyOrder(
+            "judgeAndLegalAdvisor",
+            "manageOrdersApprovalDate",
+            "orderAppliesToAllChildren",
+            "children_label",
+            "childSelector",
+            "manageOrdersEpoType",
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/ManageOrderDocumentScopedFieldsCalculatorTest.java:19]: nice that you have added it here but also add to controller test pls

File: ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json (hunk 2 above)
Comment (lewisbirks) [inline ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json:51]: This question block should be combined with the prior one as they are all the same block, this part is just a subsection of the former

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPORemovalAddressValidator.java (hunk 16 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPORemovalAddressValidator.java:33]: I added it assuming we need postcode validation. UX confirmed that we don't need postcode validation, so I'll remove this class and tests

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java (hunk 18 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java:44]: I'm using dummy daya for downloading the additional document from `DocumentReference` type. `convertToPdf`method should return the actual file data, which will then be merged using pdf box utility class

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    void shouldMergeTheDocumentsWhenAdditionalDocumentsAreEmpty() {
+        DocmosisDocument actualMergedPdf = underTest.mergeDocuments(DOCMOSIS_DOCUMENT, List.of());
+
+        assertThat(actualMergedPdf).isEqualTo(DOCMOSIS_DOCUMENT);
+    }
+
+    @Test
+    void shouldThrowExceptionWhenAdditionalDocumentIsCorrupted() {
+        final byte[] originalDocument = readBytes("documents/document1.pdf");
```
Comment (aliveni) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/generator/DocumentMergerTest.java:72]: we don't need actual data here, so I'll use dummy bytes here

**Review by lewisbirks — DISMISSED** [review]

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java [hunk 25]
```diff
@@ -215,11 +232,123 @@ void reviewShouldNotAlterCaseData() {
...
+    void epoEndDateShouldReturnErrorWhenEndDateIsNotInRangeWithApprovalDate() {
+        final LocalDateTime approvalDate = LocalDateTime.now().minusDays(5);
+
+        CaseData caseData = buildCaseData().toBuilder().manageOrdersEventData(
+            buildRemoveToAccommodationEventData(approvalDate, approvalDate.plusDays(9))).build();
+
+        AboutToStartOrSubmitCallbackResponse response = postMidEvent(caseData, "order-details");
+
+        assertThat(response.getErrors()).containsOnly("Emergency protection orders cannot last longer than 8 days");
+    }
```
Comment (Rob-Doherty) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java:255]: small point but by being more strict about the boundary of the date range, e.g. `1 day plus 1 second`,  you could improve the strength of this test, which would mean even slight regressions would be caught, like 
```
        CaseData caseData = buildCaseData().toBuilder().manageOrdersEventData(
            buildRemoveToAccommodationEventData(approvalDate, approvalDate.plusDays(8).plusSeconds(1))).build();
```

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidatorTest.java [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+        assertThat(underTest.validate(caseData)).isEqualTo(List.of("Enter a valid time"));
+    }
+
+    @Test
+    void validateEPOEndDateWhenDateIsNotInRange() {
+        final LocalDateTime approvalDate = time.now();
+        CaseData caseData = CaseData.builder()
+            .manageOrdersEventData(ManageOrdersEventData.builder()
+                .manageOrdersApprovalDateTime(approvalDate)
+                .manageOrdersEndDateTime(approvalDate.plusDays(10))
```
Comment (Rob-Doherty) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidatorTest.java:58]: again, this test could be strengthened by making the end date closer to the boundary, e.g. 1 day _plus_ 1 second

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidatorTest.java [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+            List.of("Emergency protection orders cannot last longer than 8 days"));
+    }
+
+    @Test
+    void validateEPOEndDateWhenDateIsInRange() {
+        final LocalDateTime approvalDate = time.now();
+        CaseData caseData = CaseData.builder()
+            .manageOrdersEventData(ManageOrdersEventData.builder()
+                .manageOrdersApprovalDateTime(approvalDate)
+                .manageOrdersEndDateTime(approvalDate.plusDays(1))
```
Comment (Rob-Doherty) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/EPOEndDateValidatorTest.java:72]: and this test could be strengthened by making the end date closer to the other side of the boundary, e.g. 1 day _minus_ 1 second

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java [hunk 28]
```diff
@@ -215,11 +232,123 @@ void reviewShouldNotAlterCaseData() {
...
+    void epoEndDateShouldReturnErrorWhenEndDateTimeIsMidnight() {
+        final LocalDateTime endDateTime = LocalDateTime.of(dateNow().plusDays(1), LocalTime.MIDNIGHT);
+
+        CaseData caseData = CaseData.builder().manageOrdersEventData(
+            buildRemoveToAccommodationEventData(now(), endDateTime)).build();
+
+        AboutToStartOrSubmitCallbackResponse response = postMidEvent(caseData, "order-details");
+
+        assertThat(response.getErrors()).containsOnly("Enter a valid time");
+    }
```
[the remark this answers is not in the document]
Comment (Rob-Doherty) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ManageOrdersMidEventControllerTest.java:267]: ah I see, thanks

**Review by samuelzerozero — APPROVED** [review]
